README
=======
keyleds
=======

This implements a user-level driver library for
Logitech keyboards. It is a personal development project built by
capturing Logitech's software communications to the keyboard. Plus
some experiments throwing bytes at it.

* Tested on a G410 Atlas Spectrum keyboard.
* If you test it with other keyboards, please let me know what works.

Features
--------

The project is made of two parts: a library and a basic command-line
tool.

The following features are supported:

* Device enumeration and basic information (keyboard type,
  firmware version, …).
* Inspection of device capabilities and led configuration.
* Individual key led control (set and get).
* Automatic translation between led identifiers and keycodes.
* Game-mode configuration
* Report rate configuration
* Library is written is pure C

A cli tool makes those features easily usable from any script.
Though at the moment, it only supports led control.

Partial Python3 bindings are provided.

Compiling
---------

Building keyleds requires the following dependencies to be installed:

* cmake
* libc development files
* libudev development files (recommended)
* python3 development files (optional, used for python bindings)
* cython3 (optional, used for python bindings)

The following one-liner should get you up and running on debian systems::

    sudo apt-get install build-essential cmake linux-libc-dev libudev-dev python3-dev cython

Build the project with the following command::

    cd build && cmake .. && make

This will place the ``keyledsctl`` utility in the build directory.

Dealing with device permissions
-------------------------------

By default, the keyboard device is not usable by any non-root user.
This means you may either:

* run this project (the cli tool or your own program using the lib) as root.
* configure your system to make the device accessible to other users.
  On udev-based systems, you can copy ``logitech.rules`` into
  ``/etc/udev/rules.d/90-logitech-plugdev.rules`` to automatically grant
  access to members of the ``plugdev`` group.

Using the cli tool
------------------

All features are not implemented yet, here is what works now. Listing
connected, supported devices::

    $ keyledsctl list
    /dev/hidraw1 046d:c330 [111111111111]

The number in square brackets is the USB serial number of the device.

Querying device information::

    $ keyledsctl info
    Name:           Gaming Keyboard G410
    Type:           keyboard
    Model:          c33000000000
    Serial:         xxxxxxxx
    Firmware[c330]: application U1  v1.002 r20 [active]
    Firmware[aabc]: bootloader BOT v14.000 r7
    Known features: feature version gamemode name reportrate leds led-effects
    Report rates:   [1ms] 2ms 4ms 8ms
    LED block[01]:  105 keys, max_rgb(255, 255, 255)
    LED block[40]:    2 keys, max_rgb(255, 255, 255)

Setting LED state. The subcommand takes a list of directives in the form
``key=color`` and applies them in order. Keys can be either their name,
their number or the special keyword ``all``::

    $ keyledsctl set-leds all=yellow enter=green f1=#ff00cd f2=#ff0cd

This would set all keys to be yellow, except ``enter``, ``F1`` and ``F2``, which
would be green and pink respectively. Recognized colors include all CSS color
names, and rgb values in hexa notation (ala web color). A list if recognized keys
can be obtained using ``get-leds``.

Getting LED state::

    $ keyledsctl get-leds
    A=#00dcff
    B=#00dcff
    ...
    RALT=#00dcff
    RMETA=#00dcff

Each key is output on a single line. The format is the same that is used by
``set-leds``, making it possible to save/restore LED status this way::

    # Saving a block of leds, such as keys, modes, gkeys...
    $ build/keyledsctl get-leds -b keys > savedkeys
    # Restoring the block of leds
    $ xargs <savedkeys build/keyledsctl set-leds -b keys

Setting blocked keys when game mode is enabled::

    $ keyledsctl gamemode lmeta rmeta compose

Dealing with multiple devices. Either device path or USB serial ca be used::

    $ keyledsctl list
    /dev/hidraw1 046d:c330 [111111111111]
    /dev/hidraw5 046d:c330 [222222222222]
    /dev/hidraw7 046d:c330 [333333333333]
    $ keyledsctl set-leds -d /dev/hidraw1 all=red
    $ keyledsctl set-leds -d 222222222222 all=green
    $ export KEYLEDS_DEVICE=333333333333
    $ keyledsctl set-leds all=blue

This sequence would set the three attached keyboard to turn all red, all green
and all blue respectively. Note that if both and environment variable and
a command-line option are specified, the command-line option takes precedence.

Using the API
-------------

Simply include ``keyleds.h``, and link with ``libkeyleds.a``. Most functions
are self-explanatory. Have a look at ``src/keyledsctl.c`` for examples.
Open tickets if you need help. Code documentation should come with time.

Some functions allocate structures but don't have a matching ``*_free_*``
function. When you find some, please open an issue and use stdlib's ``free``
in the meantime.

Using python bindings
---------------------

Python3 bindings are experimental and still incomplete. Pull requests welcome.
To use them, simply build the project and copy ``pykeyleds.so`` into your
python project.

Here is a sample of what works::

    >>> import pykeyleds
    >>> dev = pykeyleds.Device('/dev/hidraw1', 1)
    >>> dev.name
    'Gaming Keyboard G410'
    >>> dev.type
    'keyboard'
    >>> dev.protocol
    4
    >>> dev.version
    DeviceVersion(model=c33000000000, serial=35344708, transport=8, protocols=(
        DeviceProtocol(0, product=0xc330, version='U1 v101.2.14', active=True),
        DeviceProtocol(1, product=0xaabc, version='BOTv114.0.7', active=False)
    ))
    >>> dev.features
    (1, 3, 17698, 5, 7680, 17728, 7856, 32864, 193, 6145, 6146, 32896, 32880, 6177)
    >>> dev.leds
    {'modes': KeyBlock('modes', 0x40, nb_keys=2, color=Color(255, 255, 255)),
     'keys': KeyBlock('keys', 0x01, nb_keys=105, color=Color(255, 255, 255))}
    >>> dev.leds['keys'].get_all()
    (KeyColor(KEY_A, id=4, Color(0, 205, 255),
     KeyColor(KEY_B, id=5, Color(0, 205, 255),
     ...
     KeyColor(KEY_RALT, id=230, Color(0, 205, 255),
     KeyColor(KEY_RMETA, id=231, Color(0, 205, 255))
    >>> dev.leds['keys'].set_all_keys(pykeyleds.Color(63, 191, 127))
    >>> dev.commit_leds()

All properties are read once at first access and cached. On the other hand,
methods in the form ``get_*`` query the device at every invocation.


CHANGELOG.rst
#############
Release Notes
#############

Unreleased

Bugfixes:

- Fix G-keys mask and M/MR keys — #63, courtesy of @nickbclifford.


*****************************
1.1.1 - current release
*****************************

Released on June 15th, 2020

Misc:

- Add M-keys and MR-keys manipulation support in hardware library.

Bugfixes:

- Fix compilation using gcc 10 — #49, courtesy of @domrim.
- Fix python bindings for libkeyleds — #52.
- Fix failure on some versions of Xlib — #40.

*****************************
1.1.0
*****************************

Released on April 21st, 2019

New features:

- Add G-key manipulation support in hardware library and command-line tool.
- Added heatmap effect to track most frequently used keys.

Misc:

- Many code quality improvements.

Bugfixes:

- Fix compilation using gcc 9.
- Command line tool no longer truncate feature identifiers.
- Add workaround for g510 to account for its much slower speed when updating leds — #34.

*****************************
1.0.0
*****************************

Released on March 2nd, 2019

New features:

- Layout files for G213 and G513 have been added — #34.
- Wave effect can now be used as static color cycle. Simply pass it `length: 0` — #27.
- Configuration file now supports nested key groups — #30.
- Configuration file now supports custom color aliases — #31.
- Fill effect no longer ignores opacity and can now be used for transparency effects — #33.
- It is now possible to force-refresh key colors by sending `SIGUSR1` to keyledsd — #34.

Misc:

- Layout files now use a more compact format.
- Many tiny speed improvements.
- Program size reduced by ~20%.
- When keyboard layout is unknown, keyledsd now tries to fallback to international layout. — #29.
- Major refactor opens the way to supporting G-keys in a future release.

Bugfixes:

- Feedback effect no longer crashed with option `decay: 0` — #29.
- Logo has been added to GPro layouts — #29.
- Logs generated by keyledsd now follow systemd conventions.

*****************************
0.8.0.1
*****************************

Released on February 1st, 2019

The is a maintenance release, with many technical improvements:

- Drop Qt dependency, and use libuv and sd-bus instead (much lighter).
- Switch from C++14 to C++17.
- Simplify many datastructures and matching code.
- Fix many sign inconsistencies that could have introduced bugs.

*****************************
0.7.9
*****************************

Released on November 17th, 2018

Bugfixes:

- Add names for logo and status leds for G610, G810 and GPro — #22, #26.

*****************************
0.7.8.1
*****************************

Released on September 16th, 2018

Bugfixes:

- Fix invalid layout files.

*****************************
0.7.8
*****************************

Released on September 16th, 2018

Bugfixes:

- Skip Qt cross-compilation workaround when not building the service — #25.
- Loading an empty effect list no longer causes flicker.
- Layout files now blacklist spurious keys, which should fix US backslash issues — #22.

*****************************
0.7.7
*****************************

Released on August 29th, 2018

Bugfixes:

- Fix crash in device detection when encountering non-usb hid devices. Fixes #24.

*****************************
0.7.6
*****************************

Released on July 28th, 2018

Features:

- New demo effect included: whack-a-mole game. In default config, will start whenever
  current window includes the “mole” word.

Bugfixes:

- Fix crash of wave effect occuring when no layout is loaded.
- Provide aliases for GPro keyboards layout. Fixes #22.

Misc:

- Drop support for MMX instructions. SSE2 and AVX2 are enough.
- Compile with `-ffast-math` by default.


*****************************
0.7.5
*****************************

Released on March 22th, 2018

Features:

- New effect included: idle-dim, that dims keyboard lights when it is not used.
- [Lua API] RenderTarget now supports a `fill()` and `multiply()` method.
- [Lua API] RGBAColor now supports addition/substraction/multiplication and divisions.

*****************************
0.7.4
*****************************

Replaces failed releases 0.7.2 and 0.7.3

Released on January 28th, 2018

Bugfixes:

- Fix crash on keyboards that report an empty key block (most notably G910) — #15.
- Add missing log and badge names in G910 layout files — #17.
- No longer silently skip LUA when the dependency is missing.

*****************************
0.7.1
*****************************

Released on November 15th, 2017

Features:

- Plugin API now exposes distance and angle between keys, both in LUA and C++.

Bugfixes:

- Multimedia keys are now properly detected on G810 — see #13.
- Keys that devices fail to report are now read from layout file. This fixes
  unrecognized modifier keys on G810. — see #13.

*****************************
0.7.0
*****************************

Released on October 27th, 2017

Features:

- The LUA scripting engine is there. Start writing your LUA effects now!
- A few sample effects are included, check them out.

Bugfixes:

- Having no default profile in the configuration no longer crashed the service.

*****************************
0.6.2
*****************************

Released on October 14th, 2017

Features:

- Add AVX2 optimization for blending.

Bugfixes:

- Fix crash on G810 due to incorreclty passing actual number of keys to blend
  (instead of up-aligned target capacity, required for SSE2). — issue #10.

*****************************
0.6.1
*****************************

Released on October 8th, 2017

Features:

- Feedback effect now uses sustain+decay timings instead of straight duration.
  Key color remains still for ``sustain`` milliseconds, then fades in ``decay``
  milliseconds. The old ``duration`` configuration key is no longer recognized.

Bugfixes:

- Missing (defaulted) parameters in plugin configurarion no longer cause the
  service to crash.

*****************************
0.6.0
*****************************


libkeyleds/src/keys.c
/* Keyleds -- Gaming keyboard tool
 * Copyright (C) 2017 Julien Hartmann, juli1.hartmann@gmail.com
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <stdint.h>
#include "config.h"
#include "keyleds.h"

/** Mapping of key identifier to linux input keycode, for KEYLEDS_BLOCK_KEYS block.
 * Those values come from USB HID specification for keyboards.
 */
static const uint8_t scancode_to_keycode[256] = {
      0,  0,  0,  0, 30, 48, 46, 32, 18, 33, 34, 35, 23, 36, 37, 38,    /* 0x00 */
     50, 49, 24, 25, 16, 19, 31, 20, 22, 47, 17, 45, 21, 44,  2,  3,    /* 0x10 */
      4,  5,  6,  7,  8,  9, 10, 11, 28,  1, 14, 15, 57, 12, 13, 26,    /* 0x20 */
     27, 43, 43, 39, 40, 41, 51, 52, 53, 58, 59, 60, 61, 62, 63, 64,    /* 0x30 */
     65, 66, 67, 68, 87, 88, 99, 70,119,110,102,104,111,107,109,106,    /* 0x40 */
    105,108,103, 69, 98, 55, 74, 78, 96, 79, 80, 81, 75, 76, 77, 71,    /* 0x50 */
     72, 73, 82, 83, 86,127,116,117,183,184,185,186,187,188,189,190,    /* 0x60 */
    191,192,193,194,134,138,130,132,128,129,131,137,133,135,136,113,    /* 0x70 */
    115,114,  0,  0,  0,121,  0, 89, 93,124, 92, 94, 95,  0,  0,  0,    /* 0x80 */
    122,123, 90, 91, 85,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,    /* 0x90 */
      0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,    /* 0xa0 */
      0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,    /* 0xb0 */
      0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,    /* 0xc0 */
      0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,    /* 0xd0 */
     29, 42, 56,125, 97, 54,100,126,164,166,165,163,161,115,114,113,    /* 0xe0 */
    150,158,159,128,136,177,178,176,142,152,173,140,  0,  0,  0,  0     /* 0xf0 */
};
/* Duplicate scancodes for 43, 113, 114, 115, 128, 136 might not work
 *  - 0x31 and 0x32 both result in 43 (backslash).
 *      => Logitech seems to use 0x32, so 0x31 is killed in the table
 */

/** Mapping of linux input keycode to logitech key identifier, for KEYLEDS_BLOCK_KEYS block */
static const uint8_t keycode_to_scancode[] = {
    0x00, 0x29, 0x1e, 0x1f, 0x20, 0x21, 0x22, 0x23, /* 0x00 */
    0x24, 0x25, 0x26, 0x27, 0x2d, 0x2e, 0x2a, 0x2b, /* 0x08 */
    0x14, 0x1a, 0x08, 0x15, 0x17, 0x1c, 0x18, 0x0c, /* 0x10 */
    0x12, 0x13, 0x2f, 0x30, 0x28, 0xe0, 0x04, 0x16, /* 0x18 */
    0x07, 0x09, 0x0a, 0x0b, 0x0d, 0x0e, 0x0f, 0x33, /* 0x20 */
    0x34, 0x35, 0xe1, 0x32, 0x1d, 0x1b, 0x06, 0x19, /* 0x28 */
    0x05, 0x11, 0x10, 0x36, 0x37, 0x38, 0xe5, 0x55, /* 0x30 */
    0xe2, 0x2c, 0x39, 0x3a, 0x3b, 0x3c, 0x3d, 0x3e, /* 0x38 */
    0x3f, 0x40, 0x41, 0x42, 0x43, 0x53, 0x47, 0x5f, /* 0x40 */
    0x60, 0x61, 0x56, 0x5c, 0x5d, 0x5e, 0x57, 0x59, /* 0x48 */
    0x5a, 0x5b, 0x62, 0x63, 0x00, 0x94, 0x64, 0x44, /* 0x50 */
    0x45, 0x87, 0x92, 0x93, 0x8a, 0x88, 0x8b, 0x8c, /* 0x58 */
    0x58, 0xe4, 0x54, 0x46, 0xe6, 0x00, 0x4a, 0x52, /* 0x60 */
    0x4b, 0x50, 0x4f, 0x4d, 0x51, 0x4e, 0x49, 0x4c, /* 0x68 */
    0x00, 0xef, 0xee, 0xed, 0x66, 0x67, 0x00, 0x48, /* 0x70 */
    0x00, 0x85, 0x90, 0x91, 0x89, 0xe3, 0xe7, 0x65, /* 0x78 */
    0xf3, 0x79, 0x76, 0x7a, 0x77, 0x7c, 0x74, 0x7d, /* 0x80 */
    0xf4, 0x7b, 0x75, 0x00, 0xfb, 0x00, 0xf8, 0x00, /* 0x88 */
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xf0, 0x00, /* 0x90 */
    0xf9, 0x00, 0x00, 0x00, 0x00, 0x00, 0xf1, 0xf2, /* 0x98 */
    0x00, 0xec, 0x00, 0xeb, 0xe8, 0xea, 0xe9, 0x00, /* 0xa0 */
    0x00, 0x00, 0x00, 0x00, 0x00, 0xfa, 0x00, 0x00, /* 0xa8 */
    0xf7, 0xf5, 0xf6, 0x00, 0x00, 0x00, 0x00, 0x68, /* 0xb0 */
    0x69, 0x6a, 0x6b, 0x6c, 0x6d, 0x6e, 0x6f, 0x70, /* 0xb8 */
    0x71, 0x72, 0x73, 0x00, 0x00, 0x00, 0x00, 0x00  /* 0xc0 */
};


/** Convert a device key identifier into Linux input keycode
 * @param block Device key block.
 * @param scancode Key identifier.
 * @return Linux input keycode, or 0 on failure.
 * @remark Only KEYLEDS_BLOCK_KEYS and KEYLEDS_BLOCK_MULTIMEDIA have matching keycodes.
 */
KEYLEDS_EXPORT unsigned keyleds_translate_scancode(keyleds_block_id_t block, uint8_t scancode)
{
    if (block == KEYLEDS_BLOCK_KEYS) {
        return (unsigned)scancode_to_keycode[scancode];
    }
    if (block == KEYLEDS_BLOCK_MULTIMEDIA) {
        switch (scancode) {
            case 0xb5: return 163;  /* nextsong */
            case 0xb6: return 165;  /* previoussong */
            case 0xb7: return 166;  /* stopcd */
            case 0xcd: return 164;  /* playpause */
            case 0xe2: return 113;  /* mute */
            case 0xe9: return 114;  /* volumedown */
            case 0xea: return 115;  /* volumeup */
            default: return 0;
        }
    }
    return 0;
}


/** Convert a Linux input keycode into device key block and identifier
 * @param keycode Linux input keycode.
 * @param [out] blockptr Where to store device key block. Use `NULL` if not interested.
 * @param [out] scancodeptr Where to store key identifier. Use `NULL` if not interested.
 * @return `true` on success, `false` on error.
 * @remark Only KEYLEDS_BLOCK_KEYS and KEYLEDS_BLOCK_MULTIMEDIA have matching keycodes.
 */
KEYLEDS_EXPORT bool keyleds_translate_keycode(unsigned keycode, keyleds_block_id_t * blockptr,
                                              uint8_t * scancodeptr)
{
    keyleds_block_id_t block;
    uint8_t scancode = 0;

    switch (keycode) {
        case 113: scancode = 0xe2, block = KEYLEDS_BLOCK_MULTIMEDIA; break;  /* mute */
        case 114: scancode = 0xe9, block = KEYLEDS_BLOCK_MULTIMEDIA; break;  /* volumedown */
        case 115: scancode = 0xea, block = KEYLEDS_BLOCK_MULTIMEDIA; break;  /* volumeup */
        case 163: scancode = 0xb5, block = KEYLEDS_BLOCK_MULTIMEDIA; break;  /* nextsong */
        case 164: scancode = 0xcd, block = KEYLEDS_BLOCK_MULTIMEDIA; break;  /* playpause */
        case 165: scancode = 0xb6, block = KEYLEDS_BLOCK_MULTIMEDIA; break;  /* previoussong */
        case 166: scancode = 0xb7, block = KEYLEDS_BLOCK_MULTIMEDIA; break;  /* stopcd */
        default:
            if (keycode < sizeof(keycode_to_scancode) / sizeof(keycode_to_scancode[0])) {
                scancode = keycode_to_scancode[keycode], block = KEYLEDS_BLOCK_KEYS;
            }
    }
    if (scancode == 0) { return false; }

    if (blockptr) { *blockptr = block; }
    if (scancodeptr) { *scancodeptr = scancode; }
    return true;
}
